### ISSUE TITLE
Assertion failed in the Extend implementation

### ISSUE THREAD

**Issue by Stealth2800** [gnzlbg/slice_deque#45]
So I'm using the `rodio` crate for mp3 playback in a project I'm working on and the issue I'm currently having is the sound abruptly stopping after ~8 seconds of playback from trying to decode the next frame of the mp3 data. The issue seems to stem from `SliceDeque`'s implementation of `Extend`.

[From minimp3-rs's source:](https://github.com/germangb/minimp3-rs/blob/d0eb502c88d749453f096b76421f441b38653a80/src/lib.rs#L134-L140)
```
let mut dat: [u8; MAX_SAMPLES_PER_FRAME * 5] = [0; MAX_SAMPLES_PER_FRAME * 5];
let read_bytes = self.reader.read(&mut dat)?;
self.buffer.extend(dat.iter());
```

I'm not sure what information is relevant here other than the backtrace, but:

- OS: Windows 10 (10.0.17134.165)
- Rust version: nightly-msvc [rustc 1.29.0-nightly (6a1c0637c 2018-07-23)]
- mp3 being played from memory (1,907,543 bytes), although the panic still occurred with a few other mp3 files I tested

Backtrace:
```
thread 'rodio audio processing' panicked at 'assertion failed: self.head <= self.head_upper_bound()', C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\slice-deque-0.1.10\src\lib.rs:814:9
stack backtrace:
   0:     0x7ff66bd46b42 - std::sys::windows::backtrace::set_frames
                               at C:\projects\rust\src\libstd\sys\windows\backtrace\mod.rs:104
   1:     0x7ff66bd46b42 - std::sys::windows::backtrace::set_frames
                               at C:\projects\rust\src\libstd\sys\windows\backtrace\mod.rs:104
   2:     0x7ff66bd583ce - std::sys_common::backtrace::_print
                               at C:\projects\rust\src\libstd\sys_common\backtrace.rs:71
   3:     0x7ff66bd583ce - std::sys_common::backtrace::_print
                               at C:\projects\rust\src\libstd\sys_common\backtrace.rs:71
   4:     0x7ff66bd4a5a8 - std::panicking::default_hook::{{closure}}
                               at C:\projects\rust\src\libstd\panicking.rs:211
   5:     0x7ff66bd4a261 - std::panicking::default_hook
                               at C:\projects\rust\src\libstd\panicking.rs:227
   6:     0x7ff66bd4adac - std::panicking::rust_panic_with_hook
                               at C:\projects\rust\src\libstd\panicking.rs:475
   7:     0x7ff66bc9b722 - std::panicking::begin_panic<str*>
                               at C:\projects\rust\src\libstd\panicking.rs:409
   8:     0x7ff66b636ceb - slice_deque::SliceDeque<u8>::move_tail_unchecked<u8>
                               at \<panic macros>:3
   9:     0x7ff66b636494 - slice_deque::SliceDeque<u8>::extend_desugared<u8,core::iter::Cloned<core::slice::Iter<u8>>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\slice-deque-0.1.10\src\lib.rs:1686
  10:     0x7ff66b637086 - slice_deque::{{impl}}::spec_extend<u8,core::iter::Cloned<core::slice::Iter<u8>>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\slice-deque-0.1.10\src\lib.rs:2564
  11:     0x7ff66b635eaf - slice_deque::{{impl}}::spec_extend<u8,core::slice::Iter<u8>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\slice-deque-0.1.10\src\lib.rs:2653
  12:     0x7ff66b635e4f - slice_deque::{{impl}}::extend<u8,core::slice::Iter<u8>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\slice-deque-0.1.10\src\lib.rs:2807
  13:     0x7ff66b8673e3 - minimp3::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>::refill<std::io::cursor::Cursor<bytes::bytes::Bytes>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\minimp3-0.3.0\src\lib.rs:137
  14:     0x7ff66b866a88 - minimp3::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>::next_frame<std::io::cursor::Cursor<bytes::bytes::Bytes>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\minimp3-0.3.0\src\lib.rs:77
  15:     0x7ff66b658fb7 - rodio::decoder::mp3::{{impl}}::next<std::io::cursor::Cursor<bytes::bytes::Bytes>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\decoder\mp3.rs:68
  16:     0x7ff66b72125e - rodio::decoder::{{impl}}::next<std::io::cursor::Cursor<bytes::bytes::Bytes>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\decoder\mod.rs:119
  17:     0x7ff66b721155 - core::iter::iterator::{{impl}}::next<rodio::decoder::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>>
                               at C:\projects\rust\src\libcore\iter\iterator.rs:2565
  18:     0x7ff66b6015ff - core::iter::{{impl}}::next<mut rodio::decoder::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>*>
                               at C:\projects\rust\src\libcore\iter\mod.rs:2366
  19:     0x7ff66b4b0c96 - alloc::vec::{{impl}}::from_iter<i16,core::iter::Take<mut rodio::decoder::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>*>>
                               at C:\projects\rust\src\liballoc\vec.rs:1849
  20:     0x7ff66b4b1a1d - alloc::vec::{{impl}}::from_iter<i16,core::iter::Take<mut rodio::decoder::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>*>>
                               at C:\projects\rust\src\liballoc\vec.rs:1761
  21:     0x7ff66b600b94 - core::iter::iterator::Iterator::collect<core::iter::Take<mut rodio::decoder::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>*>,alloc::vec::Vec<i16>>
                               at C:\projects\rust\src\libcore\iter\iterator.rs:1415
  22:     0x7ff66b86a3c2 - rodio::source::buffered::extract<rodio::decoder::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\source\buffered.rs:83
  23:     0x7ff66b86a02b - rodio::source::buffered::Buffered<rodio::decoder::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>>::next_frame<rodio::decoder::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\source\buffered.rs:118
  24:     0x7ff66b86abf4 - rodio::source::buffered::{{impl}}::next<rodio::decoder::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\source\buffered.rs:159
  25:     0x7ff66b7216ae - rodio::source::repeat::{{impl}}::next<rodio::decoder::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\source\repeat.rs:41
  26:     0x7ff66b7d9bf0 - rodio::source::fadein::{{impl}}::next<rodio::source::repeat::Repeat<rodio::decoder::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\source\fadein.rs:43
  27:     0x7ff66b861799 - rodio::source::pausable::{{impl}}::next<rodio::source::fadein::FadeIn<rodio::source::repeat::Repeat<rodio::decoder::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>>>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\source\pausable.rs:61
  28:     0x7ff66b7d9f02 - rodio::source::amplify::{{impl}}::next<rodio::source::pausable::Pausable<rodio::source::fadein::FadeIn<rodio::source::repeat::Repeat<rodio::decoder::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>>>>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\source\amplify.rs:60
  29:     0x7ff66b8619d6 - rodio::source::stoppable::{{impl}}::next<rodio::source::amplify::Amplify<rodio::source::pausable::Pausable<rodio::source::fadein::FadeIn<rodio::source::repeat::Repeat<rodio::decoder::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>>>>>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\source\stoppable.rs:58
  30:     0x7ff66b861a7d - rodio::source::periodic::{{impl}}::next<rodio::source::stoppable::Stoppable<rodio::source::amplify::Amplify<rodio::source::pausable::Pausable<rodio::source::fadein::FadeIn<rodio::source::repeat::Repeat<rodio::decoder::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>>>>>>,closure>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\source\periodic.rs:57
  31:     0x7ff66b7d8452 - rodio::source::samples_converter::{{impl}}::next<rodio::source::periodic::PeriodicAccess<rodio::source::stoppable::Stoppable<rodio::source::amplify::Amplify<rodio::source::pausable::Pausable<rodio::source::fadein::FadeIn<rodio::source::repeat::Repeat<rodio::decoder::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>>>>>>, closure>,f32>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\source\samples_converter.rs:49
  32:     0x7ff66b65d912 - rodio::source::done::{{impl}}::next<rodio::source::samples_converter::SamplesConverter<rodio::source::periodic::PeriodicAccess<rodio::source::stoppable::Stoppable<rodio::source::amplify::Amplify<rodio::source::pausable::Pausable<rodio::source::fadein::FadeIn<rodio::source::repeat::Repeat<rodio::decoder::Decoder<std::io::cursor::Cursor<bytes::bytes::Bytes>>>>>>>, closure>, f32>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\source\done.rs:43
  33:     0x7ff66b8d81f7 - alloc::boxed::{{impl}}::next<Source>
                               at C:\projects\rust\src\liballoc\boxed.rs:557
  34:     0x7ff66b4eba72 - rodio::queue::{{impl}}::next<f32>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\queue.rs:172
  35:     0x7ff66b84a104 - rodio::source::uniform::{{impl}}::next<rodio::queue::SourcesQueueOutput<f32>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\source\uniform.rs:158
  36:     0x7ff66b791f7a - rodio::conversions::sample_rate::SampleRateConverter<rodio::source::uniform::Take<rodio::queue::SourcesQueueOutput<f32>>>::next_input_frame<rodio::source::uniform::Take<rodio::queue::SourcesQueueOutput<f32>>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\conversions\sample_rate.rs:107
  37:     0x7ff66b79131f - rodio::conversions::sample_rate::{{impl}}::next<rodio::source::uniform::Take<rodio::queue::SourcesQueueOutput<f32>>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\conversions\sample_rate.rs:156
  38:     0x7ff66b7d4cfb - rodio::conversions::channels::{{impl}}::next<rodio::conversions::sample_rate::SampleRateConverter<rodio::source::uniform::Take<rodio::queue::SourcesQueueOutput<f32>>>>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\conversions\channels.rs:62
  39:     0x7ff66b7de562 - rodio::conversions::sample::{{impl}}::next<rodio::conversions::channels::ChannelCountConverter<rodio::conversions::sample_rate::SampleRateConverter<rodio::source::uniform::Take<rodio::queue::SourcesQueueOutput<f32>>>>,f32>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\conversions\sample.rs:38
  40:     0x7ff66b849735 - rodio::source::uniform::{{impl}}::next<rodio::queue::SourcesQueueOutput<f32>,f32>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\source\uniform.rs:88
  41:     0x7ff66b8d81f7 - alloc::boxed::{{impl}}::next<Source>
                               at C:\projects\rust\src\liballoc\boxed.rs:557
  42:     0x7ff66b8dc2ab - rodio::dynamic_mixer::{{impl}}::next<f32>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\dynamic_mixer.rs:125
  43:     0x7ff66b8d953e - rodio::engine::audio_callback
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\engine.rs:92
  44:     0x7ff66b8d9beb - rodio::engine::play_raw::{{impl}}::deref::__static_ref_initialize::{{closure}}::{{closure}}
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\engine.rs:40
  45:     0x7ff66b8d637a - cpal::{{impl}}::run::{{closure}}<closure>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\cpal-0.8.1\src\lib.rs:489
  46:     0x7ff66b9711f5 - cpal::cpal_impl::stream::EventLoop::run_inner
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\cpal-0.8.1\src\wasapi\stream.rs:589
  47:     0x7ff66b8e36bd - cpal::cpal_impl::stream::EventLoop::run<closure>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\cpal-0.8.1\src\wasapi\stream.rs:423
  48:     0x7ff66b8d62c6 - cpal::EventLoop::run<closure>
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\cpal-0.8.1\src\lib.rs:489
  49:     0x7ff66b8d9c39 - rodio::engine::play_raw::{{impl}}::deref::__static_ref_initialize::{{closure}}
                               at C:\Users\Austin\.cargo\registry\src\github.com-1ecc6299db9ec823\rodio-0.8.0\src\engine.rs:39
  50:     0x7ff66b8daf72 - std::sys_common::backtrace::__rust_begin_short_backtrace<closure,()>
                               at C:\projects\rust\src\libstd\sys_common\backtrace.rs:136
  51:     0x7ff66b8d8ae2 - std::thread::{{impl}}::spawn::{{closure}}::{{closure}}<closure,()>
                               at C:\projects\rust\src\libstd\thread\mod.rs:409
  52:     0x7ff66b8dadf2 - std::panic::{{impl}}::call_once<(),closure>
                               at C:\projects\rust\src\libstd\panic.rs:313
  53:     0x7ff66b8db0d5 - std::panicking::try::do_call<std::panic::AssertUnwindSafe<closure>,()>
                               at C:\projects\rust\src\libstd\panicking.rs:310
  54:     0x7ff66bd5bbc1 - panic_unwind::__rust_maybe_catch_panic
                               at C:\projects\rust\src\libpanic_unwind\lib.rs:106
  55:     0x7ff66b8dafff - std::panicking::try<(),std::panic::AssertUnwindSafe<closure>>
                               at C:\projects\rust\src\libstd\panicking.rs:289
  56:     0x7ff66b8daf92 - std::panic::catch_unwind<std::panic::AssertUnwindSafe<closure>,()>
                               at C:\projects\rust\src\libstd\panic.rs:392
  57:     0x7ff66b8d8953 - std::thread::{{impl}}::spawn::{{closure}}<closure,()>
                               at C:\projects\rust\src\libstd\thread\mod.rs:408
  58:     0x7ff66b8d8dd1 - alloc::boxed::{{impl}}::call_box<(),closure>
                               at C:\projects\rust\src\liballoc\boxed.rs:640
  59:     0x7ff66bd4b371 - alloc::boxed::{{impl}}::call_once
                               at C:\projects\rust\src\liballoc\boxed.rs:650
  60:     0x7ff66bd4b371 - alloc::boxed::{{impl}}::call_once
                               at C:\projects\rust\src\liballoc\boxed.rs:650
  61:     0x7ff66bd4b371 - alloc::boxed::{{impl}}::call_once
                               at C:\projects\rust\src\liballoc\boxed.rs:650
  62:     0x7ffc87483033 - BaseThreadInitThunk
  63:     0x7ffc87991430 - RtlUserThreadStart```

**Comment by gnzlbg** [comment]
Thanks for the report, I'll try to look into this today. 

The first thing I am going to have to do, is come up with a reproducer that only uses `slice_deque`. If you are willing to help, that would be great.

**Comment by Stealth2800** [comment]
I managed to reproduce the same panic here: https://github.com/Stealth2800/slice_deque_temp

**Comment by singron** [comment]
I've noticed that truncate_front moves head beyond head_upper_bound. I.e. truncate_front happily allows head to point into the mirrored section. I'm guessing it should have some logic to reset that before returning. E.g.

```rust
if self.head > self.head_upper_bound() {
  self.head -= self.capacity()
  self.tail -= self.capacity()
}
```

**Comment by gnzlbg** [comment]
So this is fixed on master. I've published 0.1.11 with the fix, a `cargo update` should be enough for you to get the latest version. Let me know if the issue is solved in your applications.

The issue was that `truncate_{front,back}` were not updated to use the `move_{head,tail}` methods when these were introduced. These methods manage wrapping around the head and the tail when necessary, but the `truncate` methods were doing it manually. That implementation was correct initially, but once the implementation of `move_{head,tail}` changed, it became incorrect. 

Luckily there was an `assert!` somewhere checking that. In the fix I tried to make the debug builds a bit more robust by adding similar asserts to other places in the code where these things could be caught, and I've added @Stealth2800 's test to the suite. I've inspected all other places in the code that change the position of the `head`/`tail` manually (it has to be done manually in some cases where the invariants are temporarily broken) and the rest appears correct.

But please test this and let me know if there are any issues. 

Thank you both for the report and the patience, and thank you a lot @Stealth2800 for the reproducer: it made fixing this a bliss!
